### PR TITLE
[recipe, diffusion, trainer, tests] feat: add LTX-2.3 FlowGRPO v1 sync recipe

### LINKED ISSUE #389: [RFC] Tracking v1 upgrade for verl-omni v0.3.0: remaining recipe coverage
### Feature request

This is the **v0.3.0 dashboard** for migrating remaining verl-omni models and algorithms onto the **verl v1 trainer** (TransferQueue + ReplayBuffer), and for proving each migrated path is **performance-competitive** and **behavior-compatible** with the current v0 recipes.

It does **not** redesign the v1 trainer. Architecture RFCs already exist:

| Area | RFC | Status |
|------|-----|--------|
| Diffusion v1 trainer (`sync` / `separate_async`) | #249 | Done in #296, #315 |
| Omni model base + v1 sync trainer | #232 | Done in #253, #258, #312 |
| Omni v1 `colocate_async` / `separate_async` | #320 | `separate_async` done in #380; `colocate_async` still in progress (#337); related pin/lifecycle: #491, landed in #497 |
| Diffusion OPD one-step-off teacher scheduling | #498 | Done in #495 |

**Please attach the corresponding PR in the tables below** when you start or land a (model, algorithm, mode) path. Keep this issue as the single dashboard. Only PRs against `verl-project/verl-omni` count.

Cross-reference (vllm-omni rollout, not a verl-omni v1 trainer blocker): [vllm-project/vllm-omni#778](https://github.com/vllm-project/vllm-omni/issues/778)

### Motivation

v1 is the long-term trainer control plane (aligned with upstream verl): TransferQueue, ReplayBuffer, agent-loop workers, and `sync` / `colocate_async` / `separate_async` modes.

The **architecture is in**. For **verl-omni v0.3.0**, the remaining work is **flagship recipe coverage on v1 `sync`**, not new trainer architecture. Several production recipes still launch the legacy v0 trainer (`main_diffusion`). Until those v0.3.0 models have a v1 recipe, we cannot treat v1 as the release default.

v1 is still **opt-in**. Diffusion still defaults to `trainer.use_v1=false`.

### Dashboard (updated 2026-09-11)

**v0.3.0 focus:** land remaining flagship diffusion recipes on v1 `sync`. Omni async trainers and extra algorithm ports do **not** block this list.

## Done

Merged on `verl-project/verl-omni`. These paths have a first-class v1 recipe or trainer.

| Stack | Path | Mode | Recipe / entrypoint | PR |
|-------|------|------|---------------------|----|
| Diffusion trainer | v1 infra | `sync` | `python -m verl_omni.trainer.main_diffusion_v1` | #296 |
| Diffusion trainer | v1 infra | `separate_async` | same entrypoint, `trainer.v1.trainer_mode=separate_async` | #315 |
| Diffusion | SD3.5 Medium + FlowGRPO | `sync` | `examples/flowgrpo_trainer/sd35/run_sd35_medium_ocr_lora_v1.sh` | #296 |
| Diffusion | SD3.5 Medium + FlowGRPO | `separate_async` | `examples/flowgrpo_trainer/sd35/run_sd35_medium_ocr_lora_v1_separate_async.sh` | #315 |
| Diffusion | Wan2.2 + DanceGRPO | `sync` | `examples/dancegrpo_trainer/wan22/run_wan22_5b_t2v_hpsv3_v1.sh` | #466 |
| Diffusion | SD3.5 Medium + OPD (multi-teacher) | `sync` | `examples/diffusionopd_trainer/sd35/run_sd35_medium_mopd_distill_v1.sh` | #493 |
| Diffusion | OPD teacher scheduling | `separate_async` | trainer wiring on v1 `separate_async` | #495 (RFC #498) |
| Diffusion | Qwen-Image + online DPO | `sync` | `examples/dpo_trainer/qwen_image/run_qwen_image_online_dpo_lora_v1.sh` | #536 |
| Diffusion | Qwen-Image-Edit + FlowGRPO (NPU) | `sync` | `examples/flowgrpo_trainer/qwen_image_edit/run_qwen_image_edit_lora_v1_npu.sh` | #490 |
| Diffusion | MiniMax H3 + FlowGRPO | `sync` | `examples/flowgrpo_trainer/minimax_h3/run_minimax_h3_t2va_lora_v1.sh`, `run_minimax_h3_fl2va_lora_v1.sh` | #501 |
| Omni | Qwen3-Omni Thinker + GSPO | `sync` (`omni_sync`) | `python -m verl_omni.trainer.main_omni` · `examples/gspo_trainer/qwen3_omni/run_qwen3_omni_thinker_gspo_lora_v1.sh` | #258, #312 |
| Omni | Qwen3-Omni Thinker + GSPO | `separate_async` (`omni_separate_async`) | `examples/gspo_trainer/qwen3_omni/run_qwen3_omni_thinker_gspo_lora_mmk12_separate_async_v1.sh` | #380 (RFC #320) |
| Omni | Qwen3-Omni Thinker + OPD | `sync` (`omni_sync`) | `examples/gspo_trainer/qwen3_omni/run_qwen3_omni_thinker_gspo_lora_mmk12_v1_opd_npu.sh` | #375 |

Supporting merges: #488 (preserve Qwen-Image metadata in V1 TransferQueue), #513 (union colocated reward output into the v1 diffusion trainer batch), #319 (v1 rollout-train consistency monitoring), #509 (Wan2.2 DanceGRPO docs default to v1), #522 (remove v0.3.0-deprecated Qwen3-Omni legacy path).

v0 OPD extras were dropped in favor of the v1 path: #355 closed.

## In progress (claimed — do not duplicate)

Open PRs on `verl-project/verl-omni`. Comment here if you take over a stalled draft. Do **not** open a parallel PR for these rows.

### v0.3.0 recipe PRs (finish these)

| Stack | Path | Mode | Status | PR |
|-------|------|------|--------|----|
| Diffusion | Qwen-Image + FlowGRPO | `sync` | Open | #489 |
| Diffusion | Qwen-Image-Edit + FlowGRPO | `sync` | Open | #499 |
| Diffusion | BAGEL + FlowGRPO | `sync` | Open | #511 |
| Diffusion | LTX2.3 + FlowGRPO | `sync` | Open | #574 |

### Other claimed work (not a v0.3.0 recipe gap)

| Stack | Path | Mode | Status | PR |
|-------|------|------|--------|----|
| Diffusion | Qwen-Image + MixGRPO | `sync` | Draft | #512 |
| Omni | colocate-async V1 trainer | `colocate_async` | Draft | #337 (RFC #320) |
| Diffusion | hybrid rollout switching | `separate_async` | Draft | #575 |

Related v1 work tracked on other issues / PRs (not unclaimed here):

- Qwen3-Omni Thinker DAPO: #446 → #456 (merged), #502, #540, #541, #542
- Restore async omni via the vllm-omni pin: #491 (pin/lifecycle landed in #497; separate-async trainer landed in #380)
- Diffusion v1 TransferQueue checkpoint restore (`separate_async`): #470

## Need help (v0.3.0 priority)

These two recipe ports are the remaining **v0.3.0 coverage gap**. Prefer a mechanical `sync` recipe first (`main_diffusion` → `main_diffusion_v1`, `trainer.use_v1=true`, `trainer.v1.trainer_mode=sync`).

| Stack | Path | Mode | v0 starting point | Owner |
|-------|------|------|-------------------|-------|
| Diffusion | Boogu-Image + FlowGRPO | `sync` first | `examples/flowgrpo_trainer/boogu_image/` | **@RuimingMu take this**|
| Diffusion | Qwen-Image / MiniMax H3 + DiffusionNFT | `sync` first | `examples/diffusionnft_trainer/` | **@0z5a taking this** (no PR yet) |

Please claim Boogu-Image here and attach the `verl-project/verl-omni` PR. Do not open a parallel DiffusionNFT PR. Do not open a parallel LTX2.3 PR (#574).

## Non-priority (will not block v0.3.0)

Unclaimed on purpose. Do **not** pick these up until the v0.3.0 priority rows above are taken (or already have an owner). They stay on v0 for this release.

| Stack | Path | Mode | v0 starting point |
|-------|------|------|-------------------|
| Diffusion trainer | `colocate_async` | `colocate_async` | Not implemented (needs trainer work, not only a recipe) |
| Diffusion | Qwen-Image + Flow-DPPO | `sync` first | `examples/flowdppo_trainer/qwen_image/` |
| Diffusion | Qwen-Image + GRPO-Guard | `sync` first | `examples/grpoguard_trainer/qwen_image/` |
| Diffusion | SD3.5 + DiNa-LRM (DRM) | `sync` first | `examples/flowgrpo_trainer/sd35/run_sd35_medium_drm_lora.sh` |
| Diffusion | flagship `separate_async` recipes | `separate_async` | After the matching `sync` recipe lands (Qwen-Image, Wan2.2, MiniMax H3, LTX2.3, …) |

Stays on the non-v1 path **by design**:

- Omni **offline DPO** (`uses_v1_trainer` is false for `sample_source=offline` + `trainer_type=direct_preference`)
- Diffusion **offline DPO** (`examples/dpo_trainer/sd35/`)

### How to claim

1. Comment on this issue with the (model, algorithm, mode) you will take.
2. Open the PR against `verl-project/verl-omni` and mention `#389`.
3. For a recipe-only change, keep adapters / losses untouched and attach v0-vs-v1 curves or an agreed smoke when the recipe is flagship.

### PR DESCRIPTION
**Pull request by SamitHuang**
### What does this PR do?

Closes #389 (LTX-2.3 FlowGRPO v1 sync task).
  Adds the v1 synchronous recipe and infrastructure adaptation for **LTX-2.3 text-to-audio-video FlowGRPO** using **TransferQueue + ReplayBuffer**:
  - Adds `examples/flowgrpo_trainer/ltx2/run_ltx2_3_t2av_lora_v1.sh`.
  - Resolves circular import cycle in `ltx2_flow_grpo` and registers `LTX2DiffusionSingleTurnAgentLoop` in `verl_omni.agent_loop`.
  - Fixes `sampler` parameter forwarding in `LTXForwardContext` (`vllm_omni_rollout_adapter.py`).
  - Fixes non-tensor metadata forwarding (including `audio_sample_rate`) in `DiffusionAgentLoopWorkerTQ`.
  - Refactors video and audio validation logging, artifact dumping, and explicit logger termination in `PolicyGradientDiffusionTrainerV1`.
  - Adds CPU unit test coverage for `LTX2DiffusionSingleTurnAgentLoop` registration.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test

image

image

image

The v1 Trainer is -44.25s (-2.86% faster) than v0 on LTX2.3

### API and Usage Example

  Launch LTX-2.3 FlowGRPO with the V1 trainer:
  ```bash
  bash examples/flowgrpo_trainer/ltx2/run_ltx2_3_t2av_lora_v1.sh
  ```

### Design & Code Changes

  1. **V1 Recipe Script**: `examples/flowgrpo_trainer/ltx2/run_ltx2_3_t2av_lora_v1.sh` configured with `main_diffusion_v1`, `trainer.use_v1=true`, and
  `trainer.v1.trainer_mode=sync`.
  2. **Import Decoupling**: Moved `LTX2DiffusionSingleTurnAgentLoop` import to `verl_omni/agent_loop/__init__.py` to eliminate circular dependency while preserving Hydra loop
  registry.
  3. **Rollout Context**: Passed `sampler` to `LTXForwardContext` during `run_phase` in `vllm_omni_rollout_adapter.py`.
  4. **TransferQueue Audio Forwarding**: Ensured non-tensor fields like `audio_sample_rate` are retained across TransferQueue transitions in `DiffusionAgentLoopWorkerTQ`.
  5. **Video/Audio Validation & Media Dump**: Added video shape handling (`ndim == 5`), audio multiplexing (`_export_video`), and explicit `self.logger.finish()` on trainer
  completion in `trainer_base.py`.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update the documentation.
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Review by knlnguyen1802 — COMMENTED** [review]
Please resolve bot comment

**Review by knlnguyen1802 — APPROVED** [review]

**Comment by NancyFyong** [comment]
I have a question: why does the drop in the 'clap' reward look a bit strange?

**Comment by zhtmike** [comment]
yes, the reward is actually dropping.

File: verl_omni/agent_loop/diffusion_agent_loop_tq.py [hunk 1]
```diff
@@ -222,10 +222,11 @@ async def _write_trajectory_to_tq(
             field["rm_scores"] = torch.tensor([internal.reward_score], dtype=torch.float32)
 
         extra = internal.extra_fields
-        for tensor_key in extra.keys():
-            value = extra.get(tensor_key)
+        for extra_key, value in extra.items():
             if isinstance(value, torch.Tensor):
-                field[tensor_key] = value.squeeze(0) if value.dim() >= 1 and value.shape[0] == 1 else value
+                field[extra_key] = value.squeeze(0) if value.dim() >= 1 and value.shape[0] == 1 else value
+            elif extra_key != "reward_extra_info":
```
Comment (SamitHuang) [inline verl_omni/agent_loop/diffusion_agent_loop_tq.py]: This fail-open path writes every non-tensor extra field onto the TransferQueue row. `tests/agent_loop/test_diffusion_agent_loop_correctness_on_cpu.py` (`test_tq_writer_preserves_allowlisted_non_tensor_trajectory_metadata`) documents that keys like `unrelated_metadata` must stay off the wire; after this change they land as top-level TQ fields (the test only asserts nested `extra_fields`, so it will not catch the leak). `img_shapes` is also duplicated (top-level + `extra_fields_out`). Allowlist `audio_sample_rate` (and any other required scalars) instead of forwarding the rest, and add a CPU test that `audio_sample_rate` is present while `unrelated_metadata` is absent from `field`.

Comment (SamitHuang) [inline verl_omni/agent_loop/diffusion_agent_loop_tq.py]: v0 `BaseRayDiffusionTrainer._dump_generations` writes a `.pt` fallback and JSONL `video_export_error` / `output_fallback` on ffmpeg failure (covered by `tests/trainer/diffusion/test_dump_generations_video_on_cpu.py`). This v1 copy only warns and records `output: null`, so LTX T2AV dumps lose both video and audio on a single encode error. Also the `ndim == 6` / `shape[1] == 3` layout guess is the shape-guessing trap in the contributing rubric — `normalize_video_tensor` already exists. Port the v0 fallback (or call that helper) and add a v1 CPU dump test.

Comment (SamitHuang) [inline verl_omni/agent_loop/diffusion_agent_loop_tq.py]: inputs/outputs/gts/scores are reordered with `sort_idx`, but audio and `audio_sample_rate` are still read from the unsorted DataProto. Muxed mp4s will pair the wrong waveform/rate with the sorted prompts. Apply the same permutation (and CPU `.cpu()` if needed) before `_dump_generations`.

**Review by SamitHuang — COMMENTED** [review]
Reviewed the LTX-2.3 FlowGRPO v1 sync recipe against the v0 trainer dump path and the TQ allowlist tests.

The recipe + agent-loop registry split + sampler forwarding look like the right v1 enablement. Copilot's logger.finish / max_samples notes are already addressed.

Ask for before merge (inline):
1. Do not fail-open every extra_fields key onto TransferQueue; allowlist `audio_sample_rate`.
2. Port v0 video dump fallback (`.pt` + JSONL error fields) instead of dropping failed LTX artifacts.
3. Permute audio with `sort_idx` in `_log_rollout_data`.

CPU tests were not rerun in this review. GPU v0-vs-v1 timing in the PR body is not a substitute for those three contracts.

**Comment by NancyFyong** [comment]
I think we can show some examples

**Comment by SamitHuang** [comment]
> I have a question: why does the drop in the 'clap' reward look a bit strange?

ImagBind reward is increasing clearly but clap drop in the middle. I think we need to tune the weights for these two rewards to address it.

**Comment by SamitHuang** [comment]
I re-run v0 trainer and observe similar reward patterns, showing the 'clap'  reward is not caused by the v1 trainer switch

**Comment by zhtmike** [comment]
I prefer to drop the v0/v1 script if the combined reward is not increasing; add it back once it is fixed; or add a comment say it is not verified as shown in https://github.com/verl-project/verl-omni/blob/main/examples/flowgrpo_trainer/bagel/run_bagel_ocr_lora_npu.sh#L9

**Comment by SamitHuang** [comment]
Train rollout consistency looks fine, ~1e-4. I will go on verify with fl2va

image

**Comment by SamitHuang** [comment]
I re-run v0 trainer, the validtion reward converges more clear. After comparison, here are the possible causes.

1. TransferQueue Sample Key Shuffling:
  • In V0, the $N=8$ trajectories per prompt were held strictly contiguous in memory.
  • In V1, TransferQueue workers pushed asynchronously, and ReplayBuffer.sample() returned unsorted keys. Without key sorting, samples of the same prompt were split across FSDP
    micro-batches, corrupting intra-group FlowGRPO advantage scaling.
  • Fix: Add sort_diffusion_tq_keys() in diffusion_tq_batch_to_dataproto().

2. Log-Prob Micro-Batch Size:
  • V1 used log_prob_micro_batch_size_per_gpu=16, introducing attention chunking differences compared to V0's 1.
  • Fix: Set actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=1 in V1.

Now is under training

**Comment by zhtmike** [comment]
> Now is under training

bagel is failed in converging as well in v1, might be the same root cause, see https://github.com/verl-project/verl-omni/issues/561
